File: code-generator/src/main/java/io/github/potjerodekool/openapi/internal/di/AutoConfigReader.java
package io.github.potjerodekool.openapi.internal.di;

import io.github.potjerodekool.openapi.common.dependency.Bean;
import io.github.potjerodekool.openapi.internal.di.bean.AutoConfigBeanDefinition;
import io.github.potjerodekool.openapi.internal.di.bean.BeanDefinition;
import org.objectweb.asm.*;

import java.lang.annotation.Annotation;
import java.lang.reflect.Proxy;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class AutoConfigReader extends ClassVisitor {

    private final List<BeanDefinition> beanDefinitions = new ArrayList<>();


    private final Object autoConfigInstance;

    protected AutoConfigReader(final Object autoConfigInstance) {
        super(Opcodes.ASM9);
        this.autoConfigInstance = autoConfigInstance;
    }

    @Override
    public MethodVisitor visitMethod(final int access, final String name,
                                     final String descriptor,
                                     final String signature, final String[] exceptions) {
        return new AutoConfigMethodReader(api, name, descriptor, this);
    }

    void addBeanDefinition(final BeanDefinition beanDefinition) {
        this.beanDefinitions.add(beanDefinition);
    }

    public List<BeanDefinition> getBeanDefinitions() {
        return beanDefinitions;
    }

    Object getAutoConfigInstance() {
        return autoConfigInstance;
    }
}

class AutoConfigMethodReader extends MethodVisitor {

    private final AutoConfigReader autoConfigReader;

    private final String methodName;

    private final String descriptor;

    private final Map<Class<?>, Annotation> annotations = new HashMap<>();

    protected AutoConfigMethodReader(final int api,
                                     final String methodName,
                                     final String descriptor,
                                     final AutoConfigReader autoConfigReader) {
        super(api);
        this.methodName = methodName;
        this.descriptor = descriptor;
        this.autoConfigReader = autoConfigReader;
    }

    @Override
    public AnnotationVisitor visitAnnotation(final String descriptor, final boolean visible) {
        final var className = descriptorToClassName(descriptor);
        return new AutoConfigAnnotationReader(api, className, this);
    }

    private String descriptorToClassName(final String descriptor) {
        String className = descriptor.substring(1);
        className = className.substring(0, className.length() - 1);
        return className.replace('/', '.');
    }

    public void addAnnotation(final Class<?> annotationClass, final Annotation annotation) {
        this.annotations.put(annotationClass, annotation);
    }

    @Override
    public void visitEnd() {
        if (annotations.containsKey(Bean.class)) {
            try {
                final var returnType = getClass().getClassLoader().loadClass(Type.getMethodType(descriptor).getReturnType().getClassName());
                final var beanDefinition = new AutoConfigBeanDefinition(
                        autoConfigReader.getAutoConfigInstance(),
                        methodName,
                        returnType,
                        annotations
                );
                autoConfigReader.addBeanDefinition(beanDefinition);
            } catch (ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
        }
    }
}

class AutoConfigAnnotationReader extends AnnotationVisitor {

    private final Map<String, Object> attributes = new HashMap<>();
    private final String className;
    private final AutoConfigMethodReader methodReader;

    protected AutoConfigAnnotationReader(final int api,
                                         final String className,
                                         final AutoConfigMethodReader methodReader) {
        super(api);
        this.className = className;
        this.methodReader = methodReader;
    }

    @Override
    public void visit(final String name, final Object value) {
        if (value instanceof Type type) {
            try {
                final var clazz = getClass().getClassLoader().loadClass(
                        type.getClassName()
                );
                this.attributes.put(name, clazz);
            } catch (final ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
        } else {
            this.attributes.put(name, value);
        }
    }

    @Override
    public void visitEnd() {
        final var classLoader = ClassPathScanner.getClassLoader();
        final Class<?> annotationClass;

        try {
            annotationClass = classLoader.loadClass(className);
        } catch (final ClassNotFoundException e) {
            throw new RuntimeException(e);
        }

        final var annotation = (Annotation) Proxy.newProxyInstance(
                classLoader,
                new Class[]{annotationClass},
                new AnnotationInvocationHandler(
                        annotationClass,
                        this.attributes
                )
        );

        methodReader.addAnnotation(annotationClass, annotation);
    }
}

File: code-generator/src/main/kotlin/io/github/potjerodekool/openapi/internal/di/AutoConfigReader.kt
package io.github.potjerodekool.openapi.internal.di

import io.github.potjerodekool.openapi.common.dependency.Bean
import io.github.potjerodekool.openapi.internal.di.ClassPathScanner.classLoader
import io.github.potjerodekool.openapi.internal.di.bean.AutoConfigBeanDefinition
import io.github.potjerodekool.openapi.internal.di.bean.BeanDefinition
import org.objectweb.asm.*
import java.lang.reflect.Proxy
import kotlin.reflect.KClass

class AutoConfigReader(val autoConfigInstance: Any) : ClassVisitor(Opcodes.ASM9) {
    private val _beanDefinitions: MutableList<BeanDefinition> = ArrayList()

    val beanDefinitions: List<BeanDefinition>
        get() = _beanDefinitions

    override fun visitMethod(
        access: Int, name: String,
        descriptor: String,
        signature: String?, exceptions: Array<String>?
    ): MethodVisitor {
        return AutoConfigMethodReader(api, name ,descriptor, this)
    }

    fun addBeanDefinition(beanDefinition: BeanDefinition) {
        _beanDefinitions.add(beanDefinition)
    }
}

internal class AutoConfigMethodReader(
    api: Int,
    private val methodName: String,
    private val descriptor: String,
    private val autoConfigReader: AutoConfigReader
) : MethodVisitor(api) {
    private val annotations: MutableMap<KClass<*>, Annotation> = HashMap()

    override fun visitAnnotation(descriptor: String, visible: Boolean): AnnotationVisitor {
        val className = descriptorToClassName(descriptor)
        return AutoConfigAnnotationReader(api, className, this)
    }

    private fun descriptorToClassName(descriptor: String): String {
        var className = descriptor.substring(1)
        className = className.substring(0, className.length - 1)
        return className.replace('/', '.')
    }

    fun addAnnotation(annotationClass: KClass<*>, annotation: Annotation) {
        annotations[annotationClass] = annotation
    }

    override fun visitEnd() {
        if (annotations.containsKey(Bean::class)) {
            try {
                val returnType = javaClass.classLoader.loadClass(
                    Type.getMethodType(
                        descriptor
                    ).returnType.className
                ).kotlin

                val beanDefinition = AutoConfigBeanDefinition(
                    autoConfigReader.autoConfigInstance,
                    methodName,
                    returnType,
                    annotations
                )
                autoConfigReader.addBeanDefinition(beanDefinition)
            } catch (e: ClassNotFoundException) {
                throw RuntimeException(e)
            }
        }
    }
}

internal class AutoConfigAnnotationReader(
    api: Int,
    private val className: String,
    private val methodReader: AutoConfigMethodReader
) : AnnotationVisitor(api) {
    private val attributes: MutableMap<String, Any> = HashMap()

    override fun visit(name: String, value: Any) {
        if (value is Type) {
            try {
                val clazz = javaClass.classLoader.loadClass(
                    value.className
                )
                attributes[name] = clazz
            } catch (e: ClassNotFoundException) {
                throw RuntimeException(e)
            }
        } else {
            attributes[name] = value
        }
    }

    override fun visitEnd() {
        val classLoader = classLoader
        val annotationClass: Class<*>

        try {
            annotationClass = classLoader.loadClass(className)
        } catch (e: ClassNotFoundException) {
            throw RuntimeException(e)
        }

        val annotation = Proxy.newProxyInstance(
            classLoader,
            arrayOf(annotationClass),
            AnnotationInvocationHandler(
                annotationClass,
                this.attributes
            )
        ) as Annotation

        methodReader.addAnnotation(annotationClass.kotlin, annotation)
    }
}